src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp

```diff
@@ -27,115 +110,429 @@ class ParallelReplicasReadingCoordinator::Impl
+        delayed_parts.pop_front();
+    }
 
-        LOG_TRACE(log, "Request is first in partition, accepted in {} ns: {}", watch.elapsed(), request_description);
-        return {.denied = false, .mark_ranges = std::move(request.mark_ranges)};
+    String description;
+    for (const auto & part : all_parts_to_read)
+    {
+        description += part.description.describe();
+        description += fmt::format("Replicas: ({}) --- ", fmt::join(part.replicas, ","));
     }
 
-    auto & partition_reading = partition_it->second;
+    LOG_INFO(log, "Reading state is fully initialized: {}", description);
+}
 
-    PartToRead part_to_read;
-    part_to_read.range = request.block_range;
-    part_to_read.name = part_and_projection;
 
-    auto part_intersection_res = partition_reading.part_ranges.getIntersectionResult(part_to_read);
+void DefaultCoordinator::handleInitialAllRangesAnnouncement(InitialAllRangesAnnouncement announcement)
+{
+    std::lock_guard lock(mutex);
 
-    switch (part_intersection_res)
+    updateReadingState(announcement);
+    stats[announcement.replica_num].number_of_requests +=1;
+
+    ++sent_initial_requests;
+    LOG_INFO(log, "{} {}", sent_initial_requests, replicas_count);
+    if (sent_initial_requests == replicas_count)
+        finalizeReadingState();
+}
+
+void DefaultCoordinator::selectPartsAndRanges(const PartRefs & container, size_t replica_num, size_t min_number_of_marks, size_t & current_mark_size, ParallelReadResponse & response) const
+{
+    for (const auto & part : container)
     {
-        case PartSegments::IntersectionResult::REJECT:
+        if (std::find(part->replicas.begin(), part->replicas.end(), replica_num) == part->replicas.end())
+        {
+            LOG_TEST(log, "Not found part {} on replica {}", part->description.info.getPartNameV1(), replica_num);
+            continue;
+        }
+
+        if (current_mark_size >= min_number_of_marks)
         {
-            LOG_TRACE(log, "Request rejected in {} ns: {}", watch.elapsed(), request_description);
-            return {.denied = true, .mark_ranges = {}};
+            LOG_TEST(log, "Current mark size {} is bigger than min_number_marks {}", current_mark_size, min_number_of_marks);
+            continue;
         }
-        case PartSegments::IntersectionResult::EXACTLY_ONE_INTERSECTION:
+
+        if (part->description.ranges.empty())
+        {
+            LOG_TEST(log, "Part {} is already empty in reading state", part->description.info.getPartNameV1());
+            continue;
+        }
+
+        response.description.push_back({
+            .info = part->description.info,
+            .ranges = {},
+        });
+
+        while (!part->description.ranges.empty() && current_mark_size < min_number_of_marks)
         {
-            auto marks_it = partition_reading.mark_ranges_in_part.find(part_and_projection);
+            auto range = part->description.ranges.front();
+
+            if (range.getNumberOfMarks() > min_number_of_marks)
```

Comment:
why do we cap marks number based on `min_number_of_marks`?

Reply:
We don't really cap, here we just check whether we can add the whole range `[a, b)` to the set or do we have to take a bite and split this range. For example for the full scan we will have just one big range for the whole part and of course we want to parallelise it.